internals/overlord/state/notices.go

```diff
@@ -278,6 +277,24 @@ func (s *State) AddNotice(userID *uint32, noticeType NoticeType, key string, opt
 	return notice.id, nil
 }
 
+// Warnf records a warning: if it's the first Warning with this message it'll
+// be added (with its firstOccurred and lastOccurred set to the current time),
+// otherwise the existing one will have its lastOccurred updated.
+func (s *State) Warnf(template string, args ...interface{}) {
```

Comment:
It seems this function is not called anywhere. I have two questions:

1. How is a warning created? 

- Is it that in the future, if something goes wrong, the Pebble code may call this function to create a warning?
- Or is it that a charmer can create a notice with the type set to `WarningNotice`? If this is the case, currently, [only custom notice is allowed](https://github.com/canonical/pebble/blob/master/internals/daemon/api_notices.go#L201). This feature is not easy to test, I have to comment this place out and use `state.NoticeType(payload.Type)` when calling `AddNotice`, or call `Warnf` in `api_notices.go`. After testing, though, the feature works as expected, and `last-occurrence` can be updated if the same warning is issued again later.

2. Why it doesn't return a notice ID?

Asking because it seems `state.AddNotice` returns an ID. And if we want to support the feature to okay just a single warning (see my other comment), the ID will be needed I think.

Reply:
Yes, currently Pebble itself doesn't create warnings or call `Warnf` anywhere. We might add that in the future, or other users of Pebble as a library might call it. It's not the intention to expose this to charmers (as you notices, the API prevents that now) -- charms would use custom notices.

It doesn't return a notice ID as there's no need for now. And you can actually look up a notice of type warning by key (which is the textual message of the warning).

Reply:
So to test this -- you're right, how have to add `Warnf` calls to the code manually. I'll update the PR description once I've finished doing my own manual testing.